base_layer/mmr/src/balanced_binary_merkle_proof.rs

```diff
@@ -0,0 +1,88 @@
+// Copyright 2019. The Tari Project
+//
+// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+// following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use std::marker::PhantomData;
+
+use digest::Digest;
+use tari_common::DomainDigest;
+
+use crate::{common::hash_together, BalancedBinaryMerkleTree, Hash};
+
+#[derive(Debug)]
+pub struct BalancedBinaryMerkleProof<D> {
+    pub path: Vec<Hash>,
+    pub leaf_index: usize,
+    _phantom: PhantomData<D>,
+}
+
+impl<D> BalancedBinaryMerkleProof<D>
+where D: Digest + DomainDigest
+{
+    pub fn verify_consume(mut self, root: &Hash, leaf_hash: Hash) -> bool {
+        let mut computed_root = leaf_hash;
+        for sibling in self.path.iter() {
+            if self.leaf_index & 1 == 1 {
+                computed_root = hash_together::<D>(&computed_root, sibling);
+            } else {
+                computed_root = hash_together::<D>(sibling, &computed_root);
+            }
+            self.leaf_index = (self.leaf_index - 1) >> 1;
```

Comment:
Why do we have to update the leaf index while traversing a path, to verify a Merkle proof ?